examples/autobahn_client.rs

```diff
@@ -14,170 +14,82 @@
+    let mut c = c.into_connection(true);
+    let (payload, is_text) = c.receive().await?;
+    assert!(is_text);
+    let num = usize::from_str(std::str::from_utf8(&payload)?)?;
+    log::info!("{} cases to run", num);
+    Ok(num)
 }
 
-fn run_case(n: usize) -> Result<(), Box<dyn error::Error>> {
-    let addr = "127.0.0.1:9001".parse().unwrap();
-    TcpStream::connect(&addr)
-        .map_err(|e| Box::new(e) as Box<dyn error::Error>)
-        .and_then(move |socket| {
-            let resource = format!("/runCase?case={}&agent=soketto-{}", n, SOKETTO_VERSION);
-            tokio::codec::Framed::new(socket, new_client(resource))
-                .send(())
-                .map_err(|e| Box::new(e) as Box<dyn error::Error>)
-                .and_then(|framed| {
-                    framed.into_future().map_err(|(e, _)| Box::new(e) as Box<dyn error::Error>)
-                })
-                .and_then(|(response, framed)| {
-                    if response.is_none() {
-                        let e: io::Error = io::ErrorKind::ConnectionAborted.into();
-                        return Either::A(future::err(Box::new(e) as Box<dyn error::Error>))
-                    }
-                    let connection = {
-                        let codec = base::Codec::new();
-                        let mut old = framed.into_parts();
-                        let mut new = FramedParts::new(old.io, codec);
-                        new.read_buf = old.read_buf;
-                        new.write_buf = old.write_buf;
-                        let framed = Framed::from_parts(new);
-                        let mut conn = Connection::from_framed(framed, connection::Mode::Client);
-                        conn.add_extensions(old.codec.drain_extensions());
-                        conn
-                    };
-                    Either::B(future::ok(connection))
-                })
-                .and_then(|connection| {
-                    let (sink, stream) = connection.split();
-                    let sink = sink.with(|data: base::Data| {
-                        if data.is_text() {
-                            std::str::from_utf8(data.as_ref())?;
-                        }
-                        Ok(data)
-                    });
-                    stream.forward(sink)
-                        .and_then(|(_stream, mut sink)| future::poll_fn(move || sink.close()))
-                        .map_err(|e| Box::new(e) as Box<dyn error::Error>)
-                })
-        })
-        .wait()
+async fn run_case(n: usize) -> Result<(), BoxedError> {
+    log::info!("running case {}", n);
+    let resource = format!("/runCase?case={}&agent=soketto-{}", n, SOKETTO_VERSION);
+    let s = TcpStream::connect("127.0.0.1:9001").await?;
+    let mut c = new_client(s, &resource);
+    assert_matches!(c.handshake().await?, handshake::ServerResponse::Accepted {..});
+    let mut c = c.into_connection(true);
```

Comment:
Btw, is the implementation of codec in two different parts (handshakes and then connection) necessary?

Wouldn't the usage be a bit cleaner if one type did it all, something like:
```rust
c.handshake().await?
while let Some(new_msg) = c.next().await? {
    trace!("new ws message {:?}", new_msg);
}
```
Also on implementing streams: hopefully at some point we will have some sort of `for await` syntax arrive and being a stream would probably simplify the above example even further, i suppose?

Reply:
> Btw, is the implementation of codec in two different parts (handshakes and then connection) necessary?

It makes sense to me that handling of the HTTP part and the websocket framing is unrelated. As for the API, one could probably hide it behind a single type maybe just named `WebSocket`. I give it a shot.

Reply:
So, I had a closer look at this but I prefer the current approach with a distinct handshake phase. The reason is that this prevents API misuse, e.g. once you have transitioned from handshake to a websocket connection there is no going back. If a single type whose methods just receive a `&mut self` is used instead, one could attempt to perform another handshake which would be nonsense. I would like to rule this out statically.